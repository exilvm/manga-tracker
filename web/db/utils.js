const format = require('pg-format');
const {
  NUMERIC_VALUE_OUT_OF_RANGE,
  INVALID_TEXT_REPRESENTATION,
} = require('pg-error-constants');
const dblog = require('debug')('db');

function generateEqualsColumns(o, availableColumns) {
  const cols = new Set(availableColumns);
  const sqlValues = [];
  const columns = [];
  const args = [];

  Object.keys(o)
    .filter(k => o[k] !== undefined && cols.has(k))
    .forEach((k, idx) => {
      sqlValues.push(`%I=$${idx+1}`);
      columns.push(k);
      args.push(o[k]);
    });

  const sqlCols = format(sqlValues.join(','), ...columns);

  return { sqlCols, args };
}
module.exports.generateEqualsColumns = generateEqualsColumns;

function handleError(err, res) {
  if (err.code === INVALID_TEXT_REPRESENTATION) {
    dblog(err.message);
    res.status(400).json({ error: 'Invalid data type given' });
    return;
  }

  if (err.code === NUMERIC_VALUE_OUT_OF_RANGE) {
    res.status(400).json({ error: 'Number value out of range' });
    return;
  }

  console.error(err);
  res.status(500).json({ error: 'Internal server error' });
}
module.exports.handleError = handleError;
