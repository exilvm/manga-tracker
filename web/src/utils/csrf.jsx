import { createContext, useContext } from 'react';

const CSRFContext = createContext(undefined);
export const CSRFProvider = CSRFContext.Provider;

export const useCSRF = () => useContext(CSRFContext);

/**
 * Returns the useCSRF token in the props property with the key _csrf
 * @param req
 * @returns {{props: {_csrf: String}}}
 */
export const csrfProps = ({ req }) => ({
  props: { _csrf: req.csrfToken() },
});

export const csrfHeader = (csrf) => ({
  'X-CSRF-Token': csrf,
});
