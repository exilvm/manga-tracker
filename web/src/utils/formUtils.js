export const showErrorAlways = ({
  meta: { submitError, dirtySinceLastSubmit, error },
}) => !!(((submitError && !dirtySinceLastSubmit) || error));
