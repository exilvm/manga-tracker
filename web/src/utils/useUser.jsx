import { createContext, useContext, useMemo } from 'react';


const UserContext = createContext(undefined);
export const UserProvider = UserContext.Provider;

export const useUser = () => {
  const user = useContext(UserContext);
  return useMemo(() => ({
    user,
    isAuthenticated: !!user,
    isAdmin: user && user.admin,
  }), [user]);
};
