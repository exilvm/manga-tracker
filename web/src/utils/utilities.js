import { format, formatDistanceToNowStrict } from 'date-fns';
import enLocale from 'date-fns/locale/en-GB';
import throttle from 'lodash.throttle';


// eslint-disable-next-line import/prefer-default-export
export const followUnfollow = (mangaId, serviceId) => {
  const url = serviceId ? `/api/user/follows?manga_id=${mangaId}&service_id=${serviceId}` :
                           `/api/user/follows?manga_id=${mangaId}`;
  return throttle((event) => {
    const target = event.target;
    switch (target.textContent.toLowerCase()) {
      case 'follow':
        fetch(url, { credentials: 'include', method: 'put' })
            .then(res => {
              if (res.status === 200) {
                target.textContent = 'Unfollow';
              }
            });
        break;

      case 'unfollow':
        fetch(url, { credentials: 'include', method: 'delete' })
            .then(res => {
              if (res.status === 200) {
                target.textContent = 'Follow';
              }
            });
        break;

      default:
        target.textContent = 'Follow';
    }
  }, 200, { trailing: false });
};

// This seems to be faster than a custom recursive function  according to my measurements
export function jsonSerializable(value) {
  return JSON.parse(JSON.stringify(value));
}

export function defaultDateFormat(date, ifUndefined='Unknown') {
  if (!date || Number.isNaN(date.getTime())) return ifUndefined;

  format(date, 'MMM do, HH:mm', { locale: enLocale });
}

export function defaultDateDistanceToNow(date, ifUndefined='Unknown') {
  if (!date || Number.isNaN(date.getTime())) return ifUndefined;

  return formatDistanceToNowStrict(date, { addSuffix: true });
}
