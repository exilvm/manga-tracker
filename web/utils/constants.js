const csrfMissing = 'Missing CSRF token';
const isDev = process.env.NODE_ENV !== 'production';
const isTest = process.env.NODE_ENV === 'test';


module.exports = {
  csrfMissing,
  isDev,
  isTest,
};
