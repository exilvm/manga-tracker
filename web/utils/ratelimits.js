const Redis = require('ioredis');
const { RateLimiterRedis, RateLimiterMemory } = require('rate-limiter-flexible');
const ExpressBruteFlexible = require('rate-limiter-flexible/lib/ExpressBruteFlexible');

let redisArgs;
if (process.env.REDIS_URL) {
    redisArgs = [process.env.REDIS_URL, {
        enableOfflineQueue: process.env.NODE_ENV !== 'production',
        showFriendlyErrorStack: process.env.NODE_ENV !== 'production',
    }];
} else {
    redisArgs = [{
        host: process.env.REDIS_HOST,
        port: process.env.REDIS_PORT || process.env.REDIS_URL,
        enableOfflineQueue: process.env.NODE_ENV !== 'production',
        showFriendlyErrorStack: process.env.NODE_ENV !== 'production',
    }];
}
const redis = new Redis(...redisArgs);

redis.on('error', err => {
    console.error('Redis error', err);
});

const rateLimiterMemory = new RateLimiterMemory({
  points: 60, // 300 / 5 if there are 5 processes at all
  duration: 60,
});

const rateLimitOpts = {
    storeClient: redis,
    points: 300,
    duration: 60,

    execEvenly: false,
    keyPrefix: 'rlflx',
    inmemoryBlockOnConsumed: 300,
    inmemoryBlockDuration: 60,
    insuranceLimiter: rateLimiterMemory,
};

const bruteOpts = {
    freeRetries: 10,
    minWait: 1000, // 1 second
    maxWait: 50000, // 50 seconds
    lifetime: 100, // 100 seconds
    storeClient: redis,
    keyPrefix: 'brtfrc',
};

const mangadexLimiter = new RateLimiterMemory({
  points: 3,
  duration: 6,
});

module.exports.mangadexLimiter = mangadexLimiter;

const bruteforce = new ExpressBruteFlexible(
  ExpressBruteFlexible.LIMITER_TYPES.REDIS,
  bruteOpts,
);

const rateLimiterRedis = new RateLimiterRedis(rateLimitOpts);

const rateLimiter = (req, res, next) => {
    const key = req.user ? req.user.user_id : req.ip;
    const pointsToConsume = req.user ? 1 : 5;
    rateLimiterRedis.consume(key, pointsToConsume)
        .then(() => next())
        .catch(() => {
            res.status(429).send('Too Many Requests');
        });
};

module.exports.bruteforce = bruteforce;
module.exports.rateLimiter = rateLimiter;
module.exports.redis = redis;
