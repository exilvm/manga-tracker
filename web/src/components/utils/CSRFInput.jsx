import { Field } from 'react-final-form';
import { useCSRF } from '../../utils/csrf';

export default function CSRFInput() {
  const csrf = useCSRF();
  return (
    <Field
      name='_csrf'
      component='input'
      type='hidden'
      defaultValue={csrf}
    />
  );
}
