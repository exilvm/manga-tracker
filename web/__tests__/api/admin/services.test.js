import request from 'supertest';
import { csrfMissing } from '../../../utils/constants';

import { redis } from '../../../utils/ratelimits';

import { userForbidden, userUnauthorized } from '../../constants';
import initServer from '../../initServer';
import stopServer from '../../stopServer';
import { adminUser, normalUser, withUser, expectErrorMessage } from '../../utils';

let httpServer;

beforeAll(async () => {
  ({ httpServer } = await initServer());
});

beforeEach(async () => {
  await redis.flushall();
});

afterAll(async () => stopServer(httpServer));

describe('POST /api/admin/editService', () => {
  it('Returns 403 without CSRF token', async () => {
    await request(httpServer)
      .post('/api/admin/editService')
      .expect(403)
      .expect(expectErrorMessage(csrfMissing));
  });

  it('returns unauthorized without user', async () => {
    await request(httpServer)
      .post('/api/admin/editService')
      .csrf()
      .expect(401)
      .expect(expectErrorMessage(userUnauthorized));
  });

  it('returns forbidden for non admin', async () => {
    await withUser(normalUser, async () => {
      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .expect(403)
        .expect(expectErrorMessage(userForbidden));
    });
  });

  it('returns 400 for admin without data', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .expect(400)
        .expect(expectErrorMessage(undefined, 'service_id'));

      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .send({
          service_id: 3,
        })
        .expect(400)
        .expect(expectErrorMessage('No valid fields given to update'));

      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .send({
          service_name: undefined,
          disabled: undefined,
          next_update: undefined,
          service_id: 3,
        })
        .expect(400)
        .expect(expectErrorMessage('No valid fields given to update'));
    });
  });

  it('Returns 400 with invalid data', async () => {
    await withUser(adminUser, async () => {
      // service_name
      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .send({
          service_name: [1, 2, 3],
          service_id: 3,
        })
        .expect(400)
        .expect(expectErrorMessage([1, 2, 3], 'service_name'));

      // disabled
      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .send({
          disabled: '2',
          service_id: 3,
        })
        .expect(400)
        .expect(expectErrorMessage('2', 'disabled'));

      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .send({
          disabled: null,
          service_id: 3,
        })
        .expect(400)
        .expect(expectErrorMessage(null, 'disabled'));

      // next_update
      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .send({
          next_update: 'abc',
          service_id: 3,
        })
        .expect(400)
        .expect(expectErrorMessage('abc', 'next_update'));

      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .send({
          next_update: 1602954767,
          service_id: 3,
        })
        .expect(400)
        .expect(expectErrorMessage(1602954767, 'next_update'));
    });
  });

  it('returns ok when editing with valid data', async () => {
    await withUser(adminUser, async () => {
      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .send({
          next_update: new Date(1602954767000),
          service_name: "Jaimini's Box",
          disabled: true,
          service_id: 3,
        })
        .expect(200);

      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .send({
          next_update: new Date(1205794767000),
          service_id: 3,
        })
        .expect(200);

      await request(httpServer)
        .post('/api/admin/editService')
        .csrf()
        .send({
          service_name: "Jaimini's Box",
          disabled: true,
          service_id: 3,
        })
        .expect(200);
    });
  });
});
