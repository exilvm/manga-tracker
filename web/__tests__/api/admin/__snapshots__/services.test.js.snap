// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POST /api/admin/editService/:serviceId Returns 400 with invalid service config data 1`] = `
Array [
  Object {
    "location": "body",
    "msg": "Value must be a valid ISO 8601 duration",
    "param": "serviceConfig.checkInterval",
    "value": "3Y6M4DT12H30M5S",
  },
  Object {
    "location": "body",
    "msg": "Value must be a valid ISO 8601 duration",
    "param": "serviceConfig.scheduledRunInterval",
    "value": "3Y6M4DT12H30M5S",
  },
]
`;

exports[`POST /api/admin/editService/:serviceId Returns 400 with invalid service config data 2`] = `
Array [
  Object {
    "location": "body",
    "msg": "Value must be a valid ISO 8601 duration",
    "param": "serviceConfig.checkInterval",
    "value": "T12H30M5S",
  },
  Object {
    "location": "body",
    "msg": "Value must be a valid ISO 8601 duration",
    "param": "serviceConfig.scheduledRunInterval",
    "value": "T12H30M5S",
  },
]
`;

exports[`POST /api/admin/editService/:serviceId Returns 400 with invalid service config data 3`] = `
Array [
  Object {
    "location": "body",
    "msg": "scheduledRunLimit must be between 1 and 100",
    "param": "serviceConfig.scheduledRunLimit",
    "value": "a",
  },
]
`;

exports[`POST /api/admin/editService/:serviceId Returns 400 with invalid service config data 4`] = `
Array [
  Object {
    "location": "body",
    "msg": "scheduledRunLimit must be between 1 and 100",
    "param": "serviceConfig.scheduledRunLimit",
    "value": 101,
  },
]
`;

exports[`POST /api/admin/editService/:serviceId Returns 400 with invalid service config data 5`] = `
Array [
  Object {
    "location": "body",
    "msg": "scheduledRunLimit must be between 1 and 100",
    "param": "serviceConfig.scheduledRunLimit",
    "value": 0,
  },
]
`;
